File: Bwl.Imaging/Geometry/Parameters/ParametersDictionaryList.vb
Public Class ParametersDictionaryList
    Inherits List(Of Parameter)

    Public Overloads Function Add(key As String, value As String, Optional visible As Boolean = False) As Parameter
        SyncLock Me
            If key Is Nothing OrElse key.Length = 0 Then Throw New Exception("ParametersDictionary::Add(): key Is Nothing OrElse key.Length = 0")
            For Each param In Me
                If param.Key.ToLower = key.ToLower Then
                    Throw New Exception("Parameter with this key exists")
                End If
            Next
            Dim newParam As New Parameter(key, value, visible)
            Me.Add(newParam)

            Return newParam
        End SyncLock
    End Function

    Public Overloads Sub Add(addingParameter As Parameter)
        SyncLock Me
            If addingParameter Is Nothing Then Throw New Exception("addingParameter nothing")
            If addingParameter.Key Is Nothing OrElse addingParameter.Key.Length = 0 Then Throw New Exception("ParametersDictionary::Add(): addingParameter.Key Is Nothing")
            For Each param In Me
                If param.Key.ToLower = addingParameter.Key.ToLower Then
                    Throw New Exception("Parameter with this key exists")
                End If
            Next
            MyBase.Add(addingParameter)
        End SyncLock
    End Sub

    Public Overloads Sub Remove(key As String)
        SyncLock Me
            Dim items2Remove = New LinkedList(Of Parameter)
            For Each param In Me
                If param.Key.ToLower = key.ToLower Then
                    items2Remove.AddLast(param)
                End If
            Next

            For Each item2Remove In items2Remove
                MyBase.Remove(item2Remove)
            Next
        End SyncLock
    End Sub

    Public ReadOnly Property ContainsKey(key As String) As Boolean
        Get
            SyncLock Me
                For Each param In Me
                    If param.Key.ToLower = key.ToLower Then
                        Return True
                    End If
                Next

                Return False
            End SyncLock
        End Get
    End Property

    Default Public Overloads Property Item(key As String) As Parameter
        Get
            SyncLock Me
                For Each param In Me
                    If param.Key.ToLower = key.ToLower Then
                        Return param
                    End If
                Next
                Throw New Exception("ParametersDictionary::Add(): key not found")
            End SyncLock
        End Get
        Set(value As Parameter)
            SyncLock Me
                If value Is Nothing Then Throw New Exception("ParametersDictionary::Add(): value Is Nothing")
                For Each param In Me
                    If param.Key.ToLower = key.ToLower Then
                        param.Value = value.Value
                        param.Caption = value.Caption
                        param.Unit = value.Unit
                        param.AdditionalSettings = value.AdditionalSettings
                        param.Visible = value.Visible
                    End If
                Next
                Add(value)
            End SyncLock
        End Set
    End Property

    Public Property ItemValue(key As String) As String
        Get
            SyncLock Me
                For Each param In Me
                    If param.Key.ToLower = key.ToLower Then
                        Return param.Value
                    End If
                Next
                Throw New Exception("ParametersDictionary::ItemValue(): key not found")
            End SyncLock
        End Get
        Set(value As String)
            SyncLock Me
                If value Is Nothing Then Throw New Exception("ParametersDictionary::ItemValue(): value is Nothing")
                For Each param In Me
                    If param.Key.ToLower = key.ToLower Then
                        param.Value = value
                    End If
                Next
                Add(key, value)
            End SyncLock
        End Set
    End Property
End Class

File: Bwl.Imaging/Geometry/Parameters/ParametersDictionaryList.cs
using System;
using System.Collections.Generic;

namespace Bwl.Imaging
{
    public class ParametersDictionaryList : List<Parameter>
    {

        public Parameter Add(string key, string value, bool visible = false)
        {
            lock (this)
            {
                if (key is null || key.Length == 0)
                    throw new Exception("ParametersDictionary::Add(): key Is Nothing OrElse key.Length = 0");
                foreach (var @param in this)
                {
                    if ((@param.Key.ToLower() ?? "") == (key.ToLower() ?? ""))
                    {
                        throw new Exception("Parameter with this key exists");
                    }
                }
                var newParam = new Parameter(key, value, visible);
                Add(newParam);

                return newParam;
            }
        }

        public new void Add(Parameter addingParameter)
        {
            lock (this)
            {
                if (addingParameter is null)
                    throw new Exception("addingParameter nothing");
                if (addingParameter.Key is null || addingParameter.Key.Length == 0)
                    throw new Exception("ParametersDictionary::Add(): addingParameter.Key Is Nothing");
                foreach (var @param in this)
                {
                    if ((@param.Key.ToLower() ?? "") == (addingParameter.Key.ToLower() ?? ""))
                    {
                        throw new Exception("Parameter with this key exists");
                    }
                }
                base.Add(addingParameter);
            }
        }

        public void Remove(string key)
        {
            lock (this)
            {
                var items2Remove = new LinkedList<Parameter>();
                foreach (var @param in this)
                {
                    if ((@param.Key.ToLower() ?? "") == (key.ToLower() ?? ""))
                    {
                        items2Remove.AddLast(@param);
                    }
                }

                foreach (var item2Remove in items2Remove)
                    Remove(item2Remove);
            }
        }

        public bool GetContainsKey(string key)
        {
            lock (this)
            {
                foreach (var @param in this)
                {
                    if ((@param.Key.ToLower() ?? "") == (key.ToLower() ?? ""))
                    {
                        return true;
                    }
                }

                return false;
            }
        }

        public Parameter this[string key]
        {
            get
            {
                lock (this)
                {
                    foreach (var @param in this)
                    {
                        if ((@param.Key.ToLower() ?? "") == (key.ToLower() ?? ""))
                        {
                            return @param;
                        }
                    }
                    throw new Exception("ParametersDictionary::Add(): key not found");
                }
            }
            set
            {
                lock (this)
                {
                    if (value is null)
                        throw new Exception("ParametersDictionary::Add(): value Is Nothing");
                    foreach (var @param in this)
                    {
                        if ((@param.Key.ToLower() ?? "") == (key.ToLower() ?? ""))
                        {
                            @param.Value = value.Value;
                            @param.Caption = value.Caption;
                            @param.Unit = value.Unit;
                            @param.AdditionalSettings = value.AdditionalSettings;
                            @param.Visible = value.Visible;
                        }
                    }
                    Add(value);
                }
            }
        }

        public string GetItemValue(string key)
        {
            lock (this)
            {
                foreach (var @param in this)
                {
                    if ((@param.Key.ToLower() ?? "") == (key.ToLower() ?? ""))
                    {
                        return @param.Value;
                    }
                }
                throw new Exception("ParametersDictionary::ItemValue(): key not found");
            }
        }
        public void SetItemValue(string key, string value)
        {
            lock (this)
            {
                if (value is null)
                    throw new Exception("ParametersDictionary::ItemValue(): value is Nothing");
                foreach (var @param in this)
                {
                    if ((@param.Key.ToLower() ?? "") == (key.ToLower() ?? ""))
                    {
                        @param.Value = value;
                    }
                }
                Add(key, value);
            }
        }
    }
}